### PR TITLE
Balder/use single threaded executor

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe and @vekterli PR

### CODE REVIEW

File: staging_vespalib/src/vespa/vespalib/util/singleexecutor.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    void run() override;
+    void drain_tasks();
+    void run_tasks_till(uint64_t available);
+    void wait_for_room(MonitorGuard & guard);
+    uint64_t index(uint64_t counter) const {
+        return counter & (_taskLimit.load(std::memory_order_relaxed) - 1);
+    }
+
+    uint64_t numTasks() const {
+        return _wp.load(std::memory_order_relaxed) - _rp.load(std::memory_order_relaxed);
```
Comment (havardpe) [inline staging_vespalib/src/vespa/vespalib/util/singleexecutor.h]: It seems the write thread might observe tasks as run before they are actually run

File: staging_vespalib/src/vespa/vespalib/util/singleexecutor.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        producerGuard.wait(10ms);
+        _producerNeedWakeup.store(false, std::memory_order_relaxed);
+    }
+}
+
+ThreadExecutor::Stats
+SingleExecutor::getStats() {
+    uint64_t accepted = _wp.load(std::memory_order_relaxed);
+    Stats stats(_maxPending, (accepted - _lastAccepted), 0);
+    _lastAccepted = accepted;
```
Comment (havardpe) [inline staging_vespalib/src/vespa/vespalib/util/singleexecutor.cpp]: not thread safe. minor consequences

**Review by havardpe — COMMENTED** [review]
relaxed memory order is ok for things you update yourself, but becomes dangerous when looking at things updated in other threads (prefer release/acquire in that case, unless the exactness of the result is not important)

File: staging_vespalib/src/vespa/vespalib/util/singleexecutor.h (hunk 1 above)
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/util/singleexecutor.h]: Agreed, fixed.

File: staging_vespalib/src/vespa/vespalib/util/singleexecutor.cpp (hunk 2 above)
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/util/singleexecutor.cpp]: Correct, but usage is always thread safe. Will reconsider after field testing and some maturing.

**Review by vekterli — COMMENTED** [review]
:+1:

**Pull request merged**